### ISSUE TITLE
where_clauses_object_safety regression on a trait where it shouldn't trigger

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust#106247]
```rust
#![deny(where_clauses_object_safety)]

pub trait Trait {
    fn method(&self) where Self: Sync;
}
```

I bisected https://github.com/rust-lang/rust/issues/51443#issuecomment-1367143741. According to `cargo-bisect-rustc --start 2022-12-28 --end 2022-12-29 -- check` it bisects to 6a20f7df5755d8c6b68110d2d0391a7b03268e77. This is concerning because nothing in #106209 looks remotely related to `where_clause_object_safety` so I am inferring that this is an unintentional regression.

Here are the 9 PRs in the rollup:

### Definitely not it:

- #94145 &mdash; **TEST ONLY**
- #104024 &mdash; **DIAGNOSTIC TEXT ONLY**
- #104708 &mdash; **COMMENT ONLY**
- #106184 &mdash; **DOCS ONLY**
- #106205 &mdash; **DEAD CODE**

### Highly unlikely:

- #103945 &mdash; **INTERNALS OF CORE::ITER::EMPTY**
- #105347 &mdash; **MINOR DIAGNOSTIC FORMATTING TWEAK**
- #106201 &mdash; **REPR(TRANSPARENT)-RELATED DIAGNOSTICS**

### I guess that leaves:

- #105484

which only adds a *brand new* lint pass unrelated to where_clauses_object_safety, independent of any existing code. The PR is +167 -0. It seems like the only possible effect should be adding new occurrences of multiple_supertrait_upcastable, not affecting anything else.

In any case cc @nbdd0121 @compiler-errors in case you can tell what's going on.

**Comment by dtolnay** [comment]
I confirmed that the regression no longer reproduces after reverting #105484.

I have put up a revert PR in #106248 since it really seems like @nbdd0121's PR did not intentionally have anything to do with where_clauses_object_safety.

**Comment by nbdd0121** [comment]
I think the way `where_clause_object_safety` currently being implemented is problematic. It currently fires when the compiler is checking whether a trait is object-safe or not, which doesn't make sense. It should be checked at actual use site (i.e. having `is_object_safe` return yes/no/no-but-we-accidentally-allowed-it.

Is the issue serious enough for a revert? I think a forward fix is better.

**Comment by dtolnay** [comment]
Yes I think it's serious enough for a revert. If I understand your explanation correctly, the multiple_supertrait_upcastable check gets run on every trait, which means we're now reporting scary object safety breakage warnings on literally every trait with a where-clause that mentions `Self`, for traits that no one ever though of using with `dyn`. That is a quite disruptive false positive.

(Of course you're free to race to put up a forward fix before the revert gets approved.)

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23106247.20where_clauses_object_safety.20regression.20on.20a.20trait.20w.E2.80.A6)).

@rustbot label -I-prioritize +P-critical

**Comment by Nilstrieb** [comment]
Removing P-critical as the revert has landed in #106248. Should we reprioritize this?

**Comment by apiraino** [comment]
@Nilstrieb `P-` I think labels can stay attached to an issue. In fact they're just signaling the priority, not their open/solved status; therefore a closed issue can stay `P-critical` for good (and sometimes it's also useful to have a reference or a "previous case" to refer to in case similar issues are filed).

In addition, it's also useful in case critical regressions are fixed on the stable channel so `T-infra` can decide for a stable backport (i.e. a point release).

@rustbot label P-critical

**Comment by pnkfelix** [comment]
Discussed in triage meeting. Downgrading (we do  not need to revisit every week)

@rustbot label: +P-high -P-critical

**Comment by pnkfelix** [comment]
further discussion led us to realize that this should just be closed due to the revert

(a test landed, so we should not reinject the bug when the lint re-lands)
